src/Exceptions/ExceptionWithContext.php

```diff
@@ -0,0 +1,35 @@
+<?php
+
+namespace Cascade\Exceptions;
+
+/**
+ * A custom exception class which provides a context argument for later reference.
+ */
+class ExceptionWithContext extends \Exception
```

Comment:
Something feels funny about catching things called `ExceptionWithContext`, but I guess you'd more likely be catching children which extend from this class. Should it be an Abstract? Or an Interface and Default implementation of that interface, just in case someone did want to cast a net to catch all `IContextualException` instances.